ath79: add support for Arduino Yun

mans0n opened the pull request:
Arduino Yun is a microcontroller development board, based on Atmel
ATmega32u4 and Atheros AR9331.

Specifications:
- MCU: ATmega32U4
- SoC: AR9331
- RAM: DDR2 64MB
- Flash: SPI NOR 16MB
- WiFi:
  - 2.4GHz: SoC internal
- Ethernet: 1x 10/100Mbps
- USB: 1x 2.0
- MicroSD: 1x SDHC

Notes:
- Stock firmware is based on OpenWrt AA.
- The MCU can be programmed with avrdude linuxgpio.

Installation:
1.  Update U-Boot environment variables to adapt to new partition scheme.
    You'll need to access serial console via YunSerialTerminal.
```
> setenv bootcmd "run addboard; run addtty; run addparts; run addrootfs; bootm 0x9f050000 || bootm 0x9fea0000"
> setenv mtdparts "spi0.0:256k(u-boot)ro,64k(u-boot-env),15936k(firmware),64k(nvram),64k(art)ro"
> saveenv
```
2.  Boot into stock firmware normally and perform sysupgrade with
    sysupgrade image.
```
# sysupgrade -n -F /tmp/sysupgrade.bin
```

mans0n committed:
ath79: add support for Arduino Yun

Arduino Yun is a microcontroller development board, based on Atmel
ATmega32u4 and Atheros AR9331.

Specifications:
- MCU: ATmega32U4
- SoC: AR9331
- RAM: DDR2 64MB
- Flash: SPI NOR 16MB
- WiFi:
  - 2.4GHz: SoC internal
- Ethernet: 1x 10/100Mbps
- USB: 1x 2.0
- MicroSD: 1x SDHC

Notes:
- Stock firmware is based on OpenWrt AA.
- The SoC UART can be accessed only through the MCU.
  YunSerialTerminal is recommended for access to serial console.
- Stock firmware uses non-standard 250000 baudrate by default.
- The MCU can be reprogrammed from the SoC with avrdude linuxgpio.

Installation:
1.  Update U-Boot environment variables to adapt to new partition scheme.
    > setenv bootcmd "run addboard; run addtty; run addparts; run addrootfs; bootm 0x9f050000 || bootm 0x9fea0000"
    > setenv mtdparts "spi0.0:256k(u-boot)ro,64k(u-boot-env),15936k(firmware),64k(nvram),64k(art)ro"
    > saveenv
2.  Boot into stock firmware normally and perform sysupgrade with
    sysupgrade image.
    # sysupgrade -n -F /tmp/sysupgrade.bin

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
Some questions:
- Do we need to migrate ethernet iface? It seems some devices already swapped their eth0/eth1 but not implemented any migration scripts.
  https://openwrt.org/docs/guide-user/installation/ar71xx.to.ath79
- stock firmware and ar71xx used "arduino:" led names. Would it be better to rename them to "yun:" or not?
- ar71xx used bit mask for MAC address setup. Do we need to implement it in 02_network? IMO dts approach is not good enough in this case.
- My yun device does not have any label MAC address sticker. But stock firmware shows wifi ssid containing a MAC address. Is is okay to use it as a label MAC?

adrianschmutzler commented:
Hi,

patch 1: I'd recommend to send this to the openwrt-devel mailing list.
patch 2: I'd split this into three separate patches, as it's about three separate things IMO.
patch 3: I haven't looked at the code, but this one should definitely go through the list, as it will bring more (relevant) people to the discussion. I've tried to start this discussion several months ago (http://lists.infradead.org/pipermail/openwrt-devel/2019-September/018794.html), but got the impression that nobody really cares about that anyway. Thus, I did not proceed with this topic afterwards, and now where 19.07.0 is released there will be even less urge to still provide an upgrade script. Nevertheless, IMO it would still be worthwhile to have it, so feel free to send it to the list.

> My yun device does not have any label MAC address sticker. But stock firmware shows wifi ssid containing a MAC address. Is is okay to use it as a label MAC?

I assume the WiFi SSID justs contains the MAC address of the WiFi interface, so I'd say we do not really learn something from that. Thus, I wouldn't use it as label MAC address, unless both eth0 and wifi have the same MAC address (which isn't the case if I read the rest correctly).

I will have a closer look and respond on the remaining questions later.

mans0n commented:
@adrianschmutzler 
> patch 1: I'd recommend to send this to the openwrt-devel mailing list.

Good suggestion. I'll do it right away.

> patch 2: I'd split this into three separate patches, as it's about three separate things IMO.

Okay, I will.

> I've tried to start this discussion several months ago (http://lists.infradead.org/pipermail/openwrt-devel/2019-September/018794.html)

Eww, that's much more complicated than I thought. I cannot think of any good way to ensure only upgrade from ar71xx get the remedy (other than abusing wmac path). It seems to be out of the range that I can handle, so I'll just discard the patch.

Thanks for your feedback. I've opened a discussion on the forum to draw more feedback.
https://forum.openwrt.org/t/port-arduino-yun-to-ath79/52555

DanielRIOT commented:
hi @mans0n 

some more information that might make it into your PR
I did a port a while back after a reverse engineering adventure. I t will turn a Yun in to a "standard" openwrt device with an extra microcontroller as it doesnt use their Juci "web desktop thing" and Arduino's utility scripts, but AVRdude works to upload code to the local 32u4 mcu as long as one has created the SPI port with spi-gpio-custom in a startup script, or defined it in the DTS using spi-gpio and spidev.

The pin mapping is as follows ( not always clear from their schematics or setup) 
ar9331 GPIO : function
26	: bitbang SPI SS ( used in  spi-gpio-custom to create /dev/spi1.0 )
27	: bitbang SPI MOSI
8	: bitbang SPI MISO
11	: bitbang SPI SCK
21	: SPI level shifter enable
19	: AVR handsake signal ( has 10k pullup )
6	: GP6 on arduino pcb
22	: handshake and GP6 level shifter enable
23	: UART level shifter enable
18 	: AVR reset
20	: configuration button
1	: LED0 : Wireless LED (blue)
0	: LED1 : USB/System Led (White)
17	: LED6 : LAN led (yellow)
*/

I also added 2 more pinmux definitions to the ar9330.dtsi  to use gpio 11, 26 and 27 as GPIOs on any AR9331 device 


```
pinmux_extended: pinmux-extended@18040030 {
				compatible = "pinctrl-single";
				reg = <0x18040030 0x4>;

				pinctrl-single,bit-per-mux;
				pinctrl-single,register-width = <32>;
				pinctrl-single,function-mask = <0x1>;
				#pinctrl-cells = <2>;

				//Extended GPIO Function Control GPIO_FUNCTION_2 bit9 ( 0x200) disabled jumpstart and ste GPIO11 as a GPIO )
				jumpstart_disable: pinmux_jumpstart_disable {
					pinctrl-single,bits = <0x0 0x200 0x200>;
				};
				//Extended GPIO Function Control GPIO_FUNCTION_2 bit8 ( 0x200) disabled jumpstart  )
				wps_disable: pinmux_wps_disable {
					pinctrl-single,bits = <0x0 0x100 0x100>;
				};


pinmux_mdio: pinmux_mdio@180600ac {
				compatible = "pinctrl-single";
				reg = <0x180600AC 0x4>;

				pinctrl-single,bit-per-mux;
				pinctrl-single,register-width = <32>;
				pinctrl-single,function-mask = <0x1>;
				#pinctrl-cells = <2>;

				mdio_gpio: pinmux_mdio_gpio {
					pinctrl-single,bits = <0x0 0x40000 0x40000>;
				};

			};

```

[ar9331_arduino-yun dts.zip](https://github.com/openwrt/openwrt/files/4053784/ar9331_arduino-yun.dts.zip)

adrianschmutzler commented on target/linux/ath79/dts/ar9331_arduino_yun.dts:
```diff
@@ -0,0 +1,223 @@
+};
+
+&eth1 {
+	status = "okay";
```
Do you need to enable this?

mans0n replied:
IIRC it's not for eth1, but for mdio1. mdio1 node is in eth1 node.

adrianschmutzler commented on target/linux/ath79/dts/ar9331_arduino_yun.dts:
```diff
@@ -0,0 +1,223 @@
+
+	mtd-mac-address = <&art 0x1002>;
+	mtd-mac-address-increment-byte = <3>;
+	mtd-mac-address-increment = <0x8>;
```
I assume you looked for direct sources of the address already?

mans0n replied:
It's complicated. Yun uses bit mask to generate ethernet MAC address.
https://github.com/openwrt/openwrt/blob/master/target/linux/ar71xx/files/arch/mips/ath79/mach-arduino-yun.c#L89
In my case `mtd-mac-address-increment[-byte]` worked well because wifi mac was xx:xx:xx:f4:xx:xx. The eth mac will be xx:xx:xx:fc:xx:xx in this case.
But it seems not all devices have mac address with 29th bit off.
https://github.com/arduino/openwrt-arduino-packages/commit/a2daabefe61aa22c0d882b2fded1f5dcff64f803
I assume the author made this commit becuase their device had wifi mac with 29th bit on.
So adding some offset won't have the same effect of masking certain bit.
Currently I'm thinking of making a general version of macaddr_setbit_la() and use it in 02_network.
https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/functions/system.sh#L154
But I'm not sure yet if I should follow ar71xx (wlan0 might have same mac address with eth0) or lede-yun (two address must be different)

mans0n commented:
Removed eth migration script and split ar71xx patch into three commits. Still work in progress.

adrianschmutzler commented on target/linux/ath79/dts/ar9331_arduino_yun.dts:
```diff
@@ -0,0 +1,223 @@
+		compatible = "gpio-keys";
+
+		config {
+			label = "config";
```
WPS?

mans0n replied:
I haven't decided on it yet. Maybe KEY_WPS_BUTTON is not a good choice here (nor in ar71xx). Who would use WPS on a tiny dev board?
The label on the board reads "WLAN_RST", and the stock FW uses it for network reset (5 sec) and factory reset (30 sec).
KEY_RESTART is not a good choice either. The button is easily accessible, any accidental touch might cause reboot. 
I'm thinking of using another code like BTN_0 or KEY_CONFIG. Then users can choose what the button should do by writing their own script or by creating a link to an existing script file (in /etc/rc.button).

mans0n committed:
base-files: add functions to set or clear bit in MAC address

Some devices (e.g. Arduino Yun) need bitwise operations during MAC address
setup. This commit adds generalized versions of macaddr_setbit_la(), which
are helpful when manipulating a single bit in a MAC address.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
- changed config button keycode from KEY_WPS_BUTTON to BTN_0
- moved ethernet mac setup from dts to 02_network
- enabled gpio13 for LED2 pin

It would be great if I could merge the macaddr_[un]setbit() functions into single one. Any suggestion?

mans0n commented:
Umm, can we merge the ar71xx patches first? I feel the target is going to be removed soon, so I'd be happy if the patches are merged before then. Some of them need to be backported, anyway.

adrianschmutzler commented:
IMO having the target in master shouldn't matter, as it's used for backports only anyway.
If fixes are provided here, they should be backported independent of whether the target exists in master or not.

The bigger problem will be to find someone to review that stuff.

adrianschmutzler commented:
Maybe it would have been better to had split this into two PRs in the first place.

mans0n commented:
Okay, then I'll try sending the ar71xx patches to the mailing list, and leaving the ath79 patches here.

mans0n commented:
I haven't tried kernel 5.4 yet. I'll push another patch for 5.4 after run-test.

mans0n commented:
Added 5.4 patch.

mans0n commented:
Increased spi frequency to 50MHz.

mans0n commented:
Rebased due to 8d9c1087e4d0.

mans0n commented:
@ynezz Could you also review this, please?

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

mans0n commented:
Thanks for the merge!

I've opened another PR for ar71xx fix. #3188

walterav1984 commented:
Arduino Yun rev 1 comes with U-Boot 1.1.4-dirty (Apr 10 2014 - 15:12:15) which doesn't support "saveenv". Therefor after flashing a newer OpenWrt image and reset the system boots the wrong memory location and boots the older broken OpenWrt Attitude adjustment. Retyping the setenv commands at uboot prompt from the pull request fixes booting OpenWrt 21.02.1 again, but will get stuck again on next reboot...

It is therefor required to first upgrade to U-Boot 1.1.5-linino-g2d891225 which comes with "saveenv" and already the right environment variables. Using the method described [here](https://docs.arduino.cc/tutorials/yun-rev2/YunUBootReflash#reflashing-u-boot-optional) with the file "tftp/u-boot-arduino-lede.bin" from the [following OS archive](https://downloads.arduino.cc/openwrtyun/17.11/yun-go-updater-linux64.tar.gz). 

```
md5sum u-boot-arduino-lede.bin 
31f2831fb094ce6b5306f215a294d18a  u-boot-arduino-lede.bin

binwalk u-boot-arduino-lede.bin 

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
153024        0x255C0         U-Boot version string, "U-Boot 1.1.5-linino-g2d891225 (Nov 22 2017 - 09:06:12)"
153280        0x256C0         CRC32 polynomial table, big endian

```

mans0n commented:
My bad. I replaced the old bootloader with the newer one before porting OpenWrt and didn't know that saveenv command doesn't exist in the old one. Thanks for the info.
